nixos/modules/services/misc/tiddlywiki.nix

```diff
@@ -0,0 +1,82 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.tiddlywiki;
+  listenParams = concatStrings (mapAttrsToList (n: v: " '${n}=${toString v}' ") cfg.listenOptions);
+  package = pkgs.nodePackages.tiddlywiki;
+  exe = "${package}/lib/node_modules/.bin/tiddlywiki";
+
+in {
+
+  options.services.tiddlywiki = {
+
+    enable = mkEnableOption "TiddlyWiki nodejs server";
+
+    dataDir = mkOption {
+      type = types.path;
+      default = "/var/lib/tiddlywiki";
+      description = "Path to wiki folder.";
+    };
+
+    listenOptions = mkOption {
+      type = types.attrs;
+      default = {};
+      example = {
+        credentials = "../credentials.csv";
+        readers="(authenticated)";
+        port = 3456;
+      };
+      description = ''
+        Parameters passed to <literal>--listen</literal> command.
+        Refer to <link xlink:href="https://tiddlywiki.com/#WebServer"/>
+        for details on supported values.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "tiddlywiki";
+      description = "User account under which TiddlyWiki runs.";
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "tiddlywiki";
+      description = "Group under which TiddlyWiki runs.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd = {
+      services.tiddlywiki = {
+        description = "TiddlyWiki nodejs server";
+        after = [ "network.target" "systemd-tmpfiles-setup.service" ];
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          Type = "simple";
+          Restart = "on-failure";
+          User = cfg.user;
+          Group = cfg.group;
+          ExecStartPre = "-${exe} ${cfg.dataDir} --init server";
+          ExecStart = "${exe} ${cfg.dataDir} --listen ${listenParams}";
+          };
+      };
+    };
+
+    systemd.tmpfiles.rules = mkIf (cfg.user == "tiddlywiki") [
```

Comment:
It might be convenient for the user to enable this unconditionally so the service doesn't crash if the directory doesn't exist already.

Reply:
I agree in general but there is a technical difficulty. When I remove this condition tests for `configured` scenario fails

The path `/home/somebody/wiki/path` is `dataDir` on `configured` machine

    configured# [    8.067281] systemd-tmpfiles[514]: Detected unsafe path transition /home/somebody → /home/somebody/wiki during canonicalization of /home/somebody/wiki.
    [...]
    configured# [   14.512043] tiddlywiki[701]: Error: Error creating directory '/home/somebody/wiki/path'

As I understand 
 * tmpfiles creates folder `/home/somebody/wiki` owned by root
 * unsafe transition check fails so tmpfiles exits without creating '/home/somebody/wiki/path'
 * tiddlywiki initializes and cannot find the '/home/somebody/wiki/path'
 * tiddlywiki tries to create '/home/somebody/wiki/path' and fails because `home/somebody/wiki` is owned by root

I propose to make tmpfiles rule conditional but check whether use changed `dataDir`. 
We have to use tmpfiles only for `/var/lib/tiddlywiki` because of permissions. For other cases user should configure `dataDir` where configured `user` have rights to create a folder and tiddlywiki will create folder on startup.

Reply:
I'm unsure as to why `tmpfiles` would fail under this scenario. I added the following to my `configuration.nix`:
```
  systemd.tmpfiles.rules = [
    "d '/home/aaron/wiki/path' 0750 aaron users - -"
  ];
```
The directory is created, as expected:
```
~> ls -altr /home/aaron/wiki
total 34
drwxr-x---  2 aaron users   2 Jul 13 09:39 path
drwx------ 60 aaron users 127 Jul 13 09:39 ..
drwxr-xr-x  3 root  root    3 Jul 13 09:39 .
```
Can you please elaborate a bit on the directory creation with `tmpfiles` problem?

Reply:
that is very interesting

If I apply new tmpfiles rules with `sudo nixos-rebuild switch` it creates folders as in your example.

But if I switch to new config after reboot `sudo nixos-rebuild boot; shutdow -r now` it creates wiki owned by root as I see in automated tests

Reply:
Relevant statement from documentation https://www.freedesktop.org/software/systemd/man/systemd-tmpfiles.html

> Note that for all line types that result in creation of any kind of file node (i.e. f/F, d/D/v/q/Q, p, L, c/b and C) leading directories are implicitly created if needed, owned by root with an access mode of 0755. In order to create them with different modes or ownership make sure to add appropriate d lines.

Reply:
Transition rule https://github.com/systemd/systemd/blob/2d9b74ba87a3fef313f362460d5d9a4283548ff6/src/basic/fs-util.c#L663

It violates the rule when tries to create `wiki` owned by root in `/home/somebody` owned by somebody

Reply:
I'm not quite sure I follow... in my example the `wiki` directory is owned by `root:root`, which is what I would expect in this case. There is no problem. The `path` subdirectory is also as I would expect - owned by `aaron:users`, as I specified.

At the moment I'm not sure why this doesn't work after reboot, but if I had to guess I would wonder if it has to do how/when `systemd-tmpfiles` is called during boot time on NixOS...

@flokli @arianvp @nh2 @fpletz @andir @erikarvstedt Any ideas on this one?

Reply:
I've made another clean test

    systemd.tmpfiles.rules = [
      "d '/home/nazarii/a/b' 0750 nazarii users - -"
    ];

```
sudo nixos-rebuild switch
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/4fi9ls234ny7bkgb9735dlsh7iryn53y-etc-nixos.conf.drv
  /nix/store/2g3wdxxql1a33r8zjhsh1f2k3g7b0dkd-etc.drv
  /nix/store/vsvyzrv7ki66pk0ybgi11i18c4hyrb40-nixos-system-bardiuk-19.09pre184759.88c2586958f.drv
building '/nix/store/4fi9ls234ny7bkgb9735dlsh7iryn53y-etc-nixos.conf.drv'...
building '/nix/store/2g3wdxxql1a33r8zjhsh1f2k3g7b0dkd-etc.drv'...
building '/nix/store/vsvyzrv7ki66pk0ybgi11i18c4hyrb40-nixos-system-bardiuk-19.09pre184759.88c2586958f.drv'...
activating the configuration...
setting up /etc...
reloading user units for nazarii...
setting up tmpfiles
Detected unsafe path transition /home/nazarii → /home/nazarii/a during canonicalization of /home/nazarii/a.
warning: error(s) occurred while switching to the new configuration
```

```
ls -la ~/a
total 8
drwxr-xr-x  2 root    root  4096 Jul 13 16:14 .
drwx------ 48 nazarii users 4096 Jul 13 16:14 ..
```

Reply:
I've dug a bit for relevant information about unsafe transition check

security issue https://github.com/systemd/systemd/issues/7986

commit message https://github.com/systemd/systemd/commit/f14f1806e329fe92d01f15c22a384702f0cb4ae0

Reply:
I've got response from systemd that this is a security concern https://github.com/systemd/systemd/issues/13052#issuecomment-511742908  

Looks like my unit test is not representative, nixos module also should not be used to create folder deep in user home

Reply:
@nbardiuk great work! I think what you've proposed, conditionally creating the directory, is the correct solution to this problem, generally speaking.

You can use `StateDirectory` from `systemd` to provision this directory when it resides under `/var/lib`.

Reply:
I think pointing `dataDir` deep inside a users' home is out of scope of this PR - supporting something like this would need to be solved in a way more generic fashion, if desired at all.

One idea that came up during all the `PermissionsStartOnly` refactors was to use `StateDirectory` in the nixos service configuration, and expose some helper nixos option bind-mounting the "real location" to it - that might work with homedirs too - but that's just brainstorming for now and not completely thought through.

Do we need to expose `dataDir` on the first iteration of this PR? Otherwise, I'd suggest to leave it to `/var/lib/tiddlywiki` for now, as it'll simplify the module code, and home directories currently don't work anyways.

Reply:
We could then even skip all the User/Group options, use `StateDirectory`, and use `DynamicUser=yes`

Reply:
I've replaced user/group configuration with DynamicUser and replaced dataDir with StateDirectory. I agree that those settings are not essential for now